proof/refine/ARM/Ipc_R.thy

```diff
@@ -5117,21 +5130,22 @@ lemma replyPush_sym_refs_list_refs_of_replies'[wp]:
          apply (wpsimp wp: bindScReply_sym_refs_list_refs_of_replies'
                            hoare_vcg_if_lift hoare_vcg_imp_lift' hoare_vcg_all_lift)
         apply (rule_tac Q="(\<lambda>_ s. sym_refs (list_refs_of_replies' s) \<and>
-                 (\<forall>rptr scp. obj_at' (\<lambda>ko. scReply ko = Some rptr) scp s \<longrightarrow>
-                  obj_at' (\<lambda>ko. replyNext ko = Some (Head scp)) rptr s) \<and>
-                 obj_at' (\<lambda>ko. replyPrev ko = None \<and> replyNext ko = None) replyPtr s)"
+                 (\<forall>rptr scp. obj_at' (\<lambda>ko. scReply ko = Some rptr) scp s
+                              \<longrightarrow> replySCs_of s rptr = Some scp) \<and>
+                 \<not> is_reply_linked replyPtr s \<and> replySCs_of s replyPtr = None)"
```

Comment:
I might be wrong here, but do you think we can replace `obj_at' (\<lambda>ko. scReply ko = Some rptr) scp s` with `scReplies_of` as well?
If we actually need `obj_at’`, can we decompose it into `sc_at’` and a projection?

Reply:
We almost definitely could, but I've mainly focused on the reply projections so far. I don't think much has been done with the SC projections yet, so if we started to use them more I expect that we would then need to prove that various functions preserve them.

My feeling is that this would be nice to do, but I don't know if it's a priority at the moment.

Reply:
Hmm. I agree that we don’t want to go much further with SC projections at this point, but `scReplies_of` comes up a lot because of `sc_replies_relation`. So, it might work nicely if we make more use of it, and in a sense it does relate to replies as well. Having said that I am not sure how much more work it will require to do so.